File: benchmark/2026-09-17-7677581(latest)/standard/rubrics/probes/CONC-1/python/conc1.py
# CONC-1, Python. Frozen workload: sum of sin(i)*sin(i) for i in 0..N.
# N = 500000000, split into 4 fixed contiguous chunks. W workers take chunks
# c with c % W == w, so the per-chunk partial sums and the final combination
# order are IDENTICAL at W=1 and W=4 and the printed result is bit-identical.
# Worker mechanism: multiprocessing.Process (CPython standard library).
import math
import sys
from multiprocessing import Process, Pipe

N = 500000000
CHUNKS = 4
SPAN = N // CHUNKS


def chunk_sum(c):
    s = 0.0
    start = c * SPAN
    end = start + SPAN
    for i in range(start, end):
        x = math.sin(i)
        s += x * x
    return s


def worker(w, W, conn):
    out = {}
    for c in range(CHUNKS):
        if c % W == w:
            out[c] = chunk_sum(c)
    conn.send(out)
    conn.close()


def main():
    W = 1
    a = sys.argv[1:]
    for k in range(len(a)):
        if a[k] == "--workers":
            W = int(a[k + 1])
    partial = [0.0] * CHUNKS
    procs = []
    conns = []
    for w in range(W):
        pr, pw = Pipe(False)
        p = Process(target=worker, args=(w, W, pw))
        p.start()
        pw.close()
        procs.append(p)
        conns.append(pr)
    for pr in conns:
        for c, v in pr.recv().items():
            partial[c] = v
    for p in procs:
        p.join()
    total = 0.0
    for c in range(CHUNKS):
        total = total + partial[c]
    print("workers=%d result=%.10f" % (W, total))


if __name__ == "__main__":
    main()

File: benchmark/2026-09-17-7677581(latest)/standard/rubrics/probes/CONC-1/cpp/conc1.cpp
// CONC-1, C++. Worker mechanism: std::thread (<thread>, C++ standard library).
#include <cstdio>
#include <cstdlib>
#include <cmath>
#include <thread>
#include <vector>

static const long long N = 500000000LL;
static const int CHUNKS = 4;
static const long long SPAN = N / CHUNKS;

static double chunk_sum(int c) {
    double s = 0.0;
    long long start = (long long)c * SPAN;
    long long end = start + SPAN;
    for (long long i = start; i < end; i++) { double x = std::sin((double)i); s += x * x; }
    return s;
}

int main(int argc, char** argv) {
    int W = (argc > 1) ? std::atoi(argv[1]) : 1;
    std::vector<double> partial(CHUNKS, 0.0);
    std::vector<std::thread> th;
    for (int w = 0; w < W; w++) {
        th.emplace_back([w, W, &partial]() {
            for (int c = 0; c < CHUNKS; c++) if (c % W == w) partial[c] = chunk_sum(c);
        });
    }
    for (auto& t : th) t.join();
    double total = 0.0;
    for (int c = 0; c < CHUNKS; c++) total = total + partial[c];
    std::printf("workers=%d result=%.10f\n", W, total);
    return 0;
}
